crates/dyn-abi/src/coerce.rs

```diff
@@ -0,0 +1,861 @@
+    #[cfg(not(feature = "debug"))]
+    let name = "uint";
+    trace(name, move |input: &mut &str| {
+        let (s, ((), fract)) =
+            spanned((uint_integer, opt(preceded('.', cut_err(digit1))))).parse_next(input)?;
+
+        let _ = space0(input)?;
+        let units = uint_units(input)?;
+
+        let uint = if let Some(fract) = fract {
+            let x = s
+                .parse::<f64>()
+                .map_err(|e| ErrMode::from_external_error(input, ErrorKind::Verify, e))?;
+
+            // TODO: add num_traits with libm feature for supporting this
+            #[cfg(not(feature = "std"))]
+            {
+                let _ = fract;
+                return Err(ErrMode::from_external_error(
+                    input,
+                    ErrorKind::Verify,
+                    Error::FloatNoStd(x),
+                ))
+            }
+
+            #[cfg(feature = "std")]
+            {
+                if units == 0 && x.fract() != 0.0 {
+                    return Err(ErrMode::from_external_error(
+                        input,
+                        ErrorKind::Verify,
+                        Error::FractionalNotAllowed(x.fract()),
+                    ))
+                }
+
+                if fract.len() > units {
+                    return Err(ErrMode::from_external_error(
+                        input,
+                        ErrorKind::Verify,
+                        Error::TooManyDecimals(units, fract.len()),
+                    ))
+                }
+
+                U256::try_from(x * 10f64.powi(units as i32))
+                    .map_err(|e| ErrMode::from_external_error(input, ErrorKind::Verify, e))
+            }
+        } else {
+            s.parse::<U256>()
+                .map_err(|e| ErrMode::from_external_error(input, ErrorKind::Verify, e))?
+                .checked_mul(U256::from(10usize.pow(units as u32)))
+                .ok_or_else(|| {
+                    ErrMode::from_external_error(input, ErrorKind::Verify, Error::IntOverflow)
+                })
+        }?;
+
+        if uint.bit_len() > len {
+            return Err(ErrMode::from_external_error(
+                input,
+                ErrorKind::Verify,
+                Error::IntOverflow,
+            ))
+        }
+
+        Ok(uint)
+    })
+}
+
+#[inline]
+fn uint_integer(input: &mut &str) -> PResult<()> {
+    let has_prefix = matches!(
+        input.get(..2),
+        Some("0b" | "0B" | "0o" | "0O" | "0x" | "0X")
+    );
+    if has_prefix {
+        *input = &input[2..];
+        hex_digit1(input)
+    } else {
+        digit1(input)
+    }
+    .map(drop)
+}
+
+#[inline]
+fn uint_units(input: &mut &str) -> PResult<usize> {
+    alpha0(input).and_then(|s| {
+        Ok(match s {
+            "ether" => 18,
+            "gwei" | "nano" | "nanoether" => 9,
+            "" | "wei" => 0,
+            _ => return Err(ErrMode::from_error_kind(input, ErrorKind::Fail)),
+        })
+    })
+}
```

Comment:
This is super handy, lets us replace a few utilities we have in foundry

wdyt about adding more units?

Reply:
Do you have anything in mind? Solidity only supports `wei`, `gwei`, and `ether`